src/python/pants/ivy/bootstrapper.py

```diff
@@ -96,32 +91,28 @@ def _get_classpath(self, workunit_factory):
     return self._classpath
 
   def _bootstrap_ivy_classpath(self, workunit_factory, retry=True):
-    # TODO(John Sirois): Extract a ToolCache class to control the path structure:
-    # https://jira.twitter.biz/browse/DPB-283
-
     ivy_bootstrap_dir = os.path.join(self._ivy_subsystem.get_options().pants_bootstrapdir,
                                      'tools', 'jvm', 'ivy')
     digest = hashlib.sha1()
-    if os.path.isfile(self._version_or_ivyxml):
-      with open(self._version_or_ivyxml, 'rb') as fp:
+    if self._ivyxml and os.path.isfile(self._ivyxml):
+      with open(self._ivyxml, 'rb') as fp:
         digest.update(fp.read())
-    else:
-      digest.update(self._version_or_ivyxml.encode())
+    digest.update(self._version.encode())
     classpath = os.path.join(ivy_bootstrap_dir, f'{digest.hexdigest()}')
 
     if not os.path.exists(classpath):
-      ivy = self._bootstrap_ivy(os.path.join(ivy_bootstrap_dir, 'bootstrap.jar'))
-      args = ['-confs', 'default', '-cachepath', classpath]
-      if os.path.isfile(self._version_or_ivyxml):
-        args.extend(['-ivy', self._version_or_ivyxml])
-      else:
-        args.extend(['-dependency', 'org.apache.ivy', 'ivy', self._version_or_ivyxml])
-
-      try:
-        ivy.execute(args=args, workunit_factory=workunit_factory, workunit_name='ivy-bootstrap')
-      except ivy.Error as e:
-        safe_delete(classpath)
-        raise self.Error('Failed to bootstrap an ivy classpath! {}'.format(e))
+      with safe_concurrent_creation(classpath) as safe_classpath:
```

Comment:
Do we actually need ivy to bootstrap itself? Why not directly fetch the version we want to use?

Reply:
I guess because of plugins?
https://github.com/pantsbuild/pants/blob/c12fcbbd56d5f1c59e7150e402c837eb25907d88/src/python/pants/ivy/bootstrapper.py#L31-L34